Add initial relations of entities

TimvdLippe opened the pull request.

TimvdLippe committed:
Add initial relations of entities

TimvdLippe committed:
Change Id in Course to @EmbeddedId

The Id was not needed as this was merely a double definition of the edition.

jwgmeligmeyling commented on src/main/java/nl/tudelft/planningstool/database/entities/courses/CourseRelation.java:
```diff
@@ -0,0 +1,39 @@
+@EqualsAndHashCode(of = {
+        "course", "user"
+})
+public class CourseRelation implements Serializable {
```
Because of the composite key using multiple `@Id` annotations, this entity needs an [`@IdClass`](http://docs.oracle.com/javaee/5/api/javax/persistence/IdClass.html). After browsing the web for hours I finally found a [decent resource](http://www.kawoolutions.com/Technology/JPA,_Hibernate,_and_Co./JPA_Composite_Key_Variants) on how composite keys should be annotated in JPA. Id classes should apparently be Serializable. Furthermore, they should not point to your relations, but their key (be it id classes or longs). The fields do not have to be annotated, but the attribute names must match to the name in the entity and equals/hashcode should be implemented. I believe `@IdClass` would look like the following (assuming IdClass can be mixed with embedded ID's):

``` java
@Data
public static clas CourseRelationId implements Serializable {
     private CourseEdition edition;
     private Integer user;
}
```

Then on the `CourseRelation`:

``` java
@IdClass(CourseRelationId.class)
public class CourseRelation implements Serializable {
//...
```

Something similar should be implemented for the `Occurence` entity.

GijsWeterings replied:
> for hours

The dedication is real! Thanks for figuring this one out. That website tho... welcome in 1995

jwgmeligmeyling replied:
Spread over weeks :wink: We already stumbled upon this when we seemed forced to use `@MapsId` in Moodcat :smile:

TimvdLippe committed:
Replace composite keys with @IdClass of inner classes

TimvdLippe committed:
Add broken bootstrapper.

This commit is broken, but is used to share the files with other developers.

TimvdLippe committed:
Fix broken bootstrapper.

The reason a mock was injected in the test was because @Inject was missing
from the constructor of CourseDAO.

TimvdLippe committed:
Add UserCreationTest

TimvdLippe committed:
Add AssignmentCreationTest

TimvdLippe commented:
Most of the tests are written. TODO: Occurrence and CourseRelation tests.

TimvdLippe committed:
Rename Occurence to Occurrence

TimvdLippe committed:
Test Occurrence relations with User and Assignment

TimvdLippe committed:
Add CourseRelationCreationTest

TimvdLippe committed:
Let multiple occurrences be planned for the same assignment

GijsWeterings commented:
LGTM, on travis OK we can merge
